src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,190 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head = new Node<>(null, null, null);
+    private Node<T> tail = new Node<>(null, null, null);
+    private int size = 0;
+
+    public MyLinkedList() {
+        head.next = tail;
+        tail.prev = head;
+    }
+
     @Override
     public void add(T value) {
+        Node<T> nodeToAdd = new Node<>(value, null, null);
+        if (head.data == null) {
+            head.data = value;
+            size++;
+            return;
+        }
+        if (size == 1) {
+            tail.data = value;
+            size++;
+            return;
+        }
+        Node<T> currentNode = tail;
+        tail = nodeToAdd;
+        tail.prev = currentNode;
+        currentNode.next = tail;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("invalid index");
+        }
+        if (index == size || head == null) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            linkBefore(value, head);
+            size++;
+            return;
+        }
+        Node<T> target = getNode(index);
+        linkBefore(value, target);
+        size++;
+    }
+
+    private void linkBefore(T value, Node<T> target) {
+        Node<T> newNode = new Node<>(value, null, null);
+        if (target == head) {
+            newNode.next = target;
+            target.prev = newNode;
+            head = newNode;
+            return;
+        }
+        Node<T> previousNode = target.prev;
+        newNode.next = target;
+        target.prev = newNode;
+        previousNode.next = newNode;
+        newNode.prev = previousNode;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
```

Comment:
[Common mistakes](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#dont-iterate-your-nodes-from-the-head-to-tail-or-vice-versa)
Don’t iterate your nodes from the head to tail or vice versa.